src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs

```diff
@@ -36,14 +37,17 @@ private static BoundObjectInitializerExpressionBase UpdateInitializers(BoundObje
             Debug.Assert(!_inExpressionLambda);
             Debug.Assert(rewrittenReceiver != null);
 
-            switch (initializerExpression.Kind)
+            switch (initializerExpression)
             {
-                case BoundKind.ObjectInitializerExpression:
-                    AddObjectInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenReceiver, ((BoundObjectInitializerExpression)initializerExpression).Initializers);
+                case BoundObjectInitializerExpression objectInitializer:
+                    AddObjectInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenReceiver, objectInitializer.Initializers);
                     return;
 
-                case BoundKind.CollectionInitializerExpression:
-                    AddCollectionInitializers(ref dynamicSiteInitializers, result, rewrittenReceiver, ((BoundCollectionInitializerExpression)initializerExpression).Initializers);
+                case BoundCollectionInitializerExpression collectionInitializer:
+                    var placeholder = collectionInitializer.PlaceholderOpt;
+                    AddPlaceholderReplacement(placeholder, rewrittenReceiver);
+                    AddCollectionInitializers(ref dynamicSiteInitializers, result, rewrittenReceiver, collectionInitializer.Initializers);
```

Comment:
>rewrittenReceiver [](start = 83, length = 17)

Consider passing down the placeholder instead and relying on the placeholder replacement for all scenarios not just for extension methods

Reply:
It is possible we don't even need passing the placeholder.

---
In reply to: [325394891](https://github.com/dotnet/roslyn/pull/38732#discussion_r325394891) [](ancestors = 325394891)

Reply:
There are a number of places where the rewritten receiver are used in the generated code and where there is no placeholder instance on the initial bound node. (See dynamic cases in `AddObjectInitializer` for instance.) We could pass the placeholder and look up the replacement at those locations but it seems easier to just pass the replacement directly.

---
In reply to: [325395424](https://github.com/dotnet/roslyn/pull/38732#discussion_r325395424) [](ancestors = 325395424,325394891)